File: core/kotlin-stub/source/jetbrains/mps/persistence/kotlin/KotlinClassesPersistence.java
/*
 * Copyright 2000-2021 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license that can be found in the LICENSE file.
 */
package jetbrains.mps.persistence.kotlin;

import jetbrains.mps.components.CoreComponent;
import jetbrains.mps.kotlin.stubs.common.KtFunctionNodeId;
import jetbrains.mps.kotlin.stubs.common.KtFunctionNodeId.Factory;
import jetbrains.mps.kotlin.stubs.commonStubs.KotlinCommonStubModelRootFactory;
import jetbrains.mps.kotlin.stubs.common.KotlinJvmPackageModelId;
import jetbrains.mps.kotlin.stubs.common.KotlinLanguage.ModelKind;
import jetbrains.mps.kotlin.stubs.common.KotlinPackageModelId;
import jetbrains.mps.kotlin.stubs.jvmStubs.KotlinJvmStubModelRootFactory;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.mps.openapi.persistence.PersistenceFacade;

public class KotlinClassesPersistence implements CoreComponent {
  private final PersistenceFacade myFacade;

  public KotlinClassesPersistence(@NotNull PersistenceFacade persistenceFacade) {
    myFacade = persistenceFacade;
  }

  @Override
  public void init() {
    // Kotlin common
    myFacade.setModelIdFactory(ModelKind.COMMON.name, new KotlinPackageModelId.Factory());
    myFacade.setModelRootFactory(ModelKind.COMMON.rootName, new KotlinCommonStubModelRootFactory());

    // Kotlin/JVM
    myFacade.setModelIdFactory(ModelKind.JVM.name, new KotlinJvmPackageModelId.Factory());
    myFacade.setModelRootFactory(ModelKind.JVM.rootName, new KotlinJvmStubModelRootFactory());

    // Kotlin functions node ids
    myFacade.setNodeIdFactory("ktfun", new Factory());
  }

  @Override
  public void dispose() {
    for (var value : ModelKind.values()) {
      myFacade.setModelIdFactory(value.name, null);
      myFacade.setModelRootFactory(value.rootName, null);
    }
  }
}

File: core/kotlin-stub/source/jetbrains/mps/persistence/kotlin/KotlinClassesPersistence.kt
/*
 * Copyright 2000-2021 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license that can be found in the LICENSE file.
 */
package jetbrains.mps.persistence.kotlin

import jetbrains.mps.components.CoreComponent
import jetbrains.mps.kotlin.stubs.loading.ids.migration.KtFunctionNodeId
import jetbrains.mps.kotlin.stubs.loading.KotlinStubModelRootFactory
import jetbrains.mps.kotlin.stubs.loading.kind.KotlinJvmModelKind
import jetbrains.mps.kotlin.stubs.loading.kind.KotlinModelKind
import org.jetbrains.mps.openapi.persistence.PersistenceFacade

private const val LEGACY_COMMON_ROOT_FACTORY = "kotlin_common"
private const val LEGACY_JVM_ROOT_FACTORY = "kotlin_jvm"
private const val LEGACY_JVM_MODEL_ID = "kotlinJvm"
private const val FUNCTION_NODE_ID_FACTORY = "ktfun"

class KotlinClassesPersistence(private val myFacade: PersistenceFacade) : CoreComponent {
    override fun init() {
        // Platform models
        KotlinModelKind.supported.forEach {
            myFacade.setModelIdFactory(it.modelKind, it.factory)
        }

        val factory = KotlinStubModelRootFactory()
        myFacade.setModelRootFactory(KotlinStubModelRootFactory.rootName, factory)

        // Kotlin functions node ids
        myFacade.setNodeIdFactory(FUNCTION_NODE_ID_FACTORY, KtFunctionNodeId.factory)

        // Legacy kotlin name and Kotlin/JVM specific loading (deprecated)
        myFacade.setModelIdFactory(LEGACY_JVM_MODEL_ID, KotlinJvmModelKind.factory)
        myFacade.setModelRootFactory(LEGACY_COMMON_ROOT_FACTORY, factory)
        myFacade.setModelRootFactory(LEGACY_JVM_ROOT_FACTORY, factory)
    }

    override fun dispose() {
        KotlinModelKind.supported.forEach {
            myFacade.setModelIdFactory(it.modelKind, null)
        }

        myFacade.setModelRootFactory(KotlinStubModelRootFactory.rootName, null)
        myFacade.setNodeIdFactory(FUNCTION_NODE_ID_FACTORY, null)

        // Legacy
        myFacade.setModelIdFactory(LEGACY_JVM_MODEL_ID, null)
        myFacade.setModelRootFactory(LEGACY_COMMON_ROOT_FACTORY, null)
        myFacade.setModelRootFactory(LEGACY_JVM_ROOT_FACTORY, null)
    }
}
